### PR TITLE
extend githubactions for trello integration 

### PR DESCRIPTION
**Pull request by lfbdev**
# Summary

new githubactions for trello integration 

## Key Points

- [ ] This is a breaking change.
- [ ] Documentation (new or existing) is updated.

### Current Behavior

Describe the current behavior of this issue, if relevant.

### New Behavior

Describe the newly updated behavior, if relevant.

### Screenshots

Include any relevant screenshots here.

### Other Information

Any other information that is important to this PR.

### CODE REVIEW

File: examples/config.yaml [hunk 1]
```diff
@@ -32,6 +32,16 @@ tools:
       name: go
       version: "1.17"
     branch: main
+- name: trello-integration
```
Comment (daniel-hutao) [inline examples/config.yaml:35]: Maybe githubintegrations-trello is better.

File: internal/pkg/plugin/githubintegrations/githubintegrations.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+	"bytes"
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"strings"
+	"text/template"
+
+	"github.com/spf13/viper"
+
```
Comment (daniel-hutao) [inline internal/pkg/plugin/githubintegrations/githubintegrations.go:13]: blank lines here

File: internal/pkg/plugin/githubintegrations/githubintegrations.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
+package githubintegrations
```
Comment (daniel-hutao) [inline internal/pkg/plugin/githubintegrations/githubintegrations.go:1]: It's very similar to githubactions, can we reuse the code by some way.

**Review by daniel-hutao — CHANGES_REQUESTED** [review]
from me

File: internal/pkg/plugin/githubintegrations/githubintegrations.go (hunk 3 above)
Comment (IronCore864) [reply internal/pkg/plugin/githubintegrations/githubintegrations.go:1]: I agree with Daniel here. I think the name should at least have "trello" in it. Like "trellogithub" or something like that.

File: internal/pkg/plugin/githubintegrations/githubintegrations.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+	tc := oauth2.NewClient(ctx, ts)
+	return github.NewClient(tc), nil
+}
+
+func verifyOptions(opt *Options) bool {
+	return opt.Owner != "" &&
+		opt.Repo != "" &&
+		opt.Branch != "" &&
+		opt.Api != nil
+}
```
Comment (IronCore864) [inline internal/pkg/plugin/githubintegrations/githubintegrations.go:254]: This file is too long; I think we can split it up.
For example, non-exported functions can be put in another file.
For another, `VerifyWorkflows` is too long; I must scroll up and down to get an overview of it. Maybe we can extract some functions, like `checkResponseCode`, `checkAllWorkflowYAMLs`, etc. The name is only an example; I didn't think it though.

File: internal/pkg/plugin/githubintegrations/githubintegrations.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+// renderTemplate render the github actions template with config.yaml
+func (gi *GithubIntegrations) renderTemplate(workflow *Workflow) error {
+	var jobs Jobs
+	err := mapstructure.Decode(gi.options.Jobs, &jobs)
+	if err != nil {
+		return err
+	}
+	//if use default {{.}}, it will confict (github actions vars also use them)
+	t, err := template.New("githubactions").Delims("[[", "]]").Parse(workflow.workflowContent)
```
Comment (IronCore864) [inline internal/pkg/plugin/githubintegrations/githubintegrations.go:219]: it's not githubactions any more

File: internal/pkg/plugin/githubintegrations/githubintegrations.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+	// error reason is 404
+	if resp.StatusCode == http.StatusNotFound {
+		return "", nil
+	}
+
+	// no error occurred
+	if resp.StatusCode == http.StatusOK {
+		return *content.SHA, nil
+	}
+	return "", fmt.Errorf("got some error is not expected")
```
Comment (IronCore864) [inline internal/pkg/plugin/githubintegrations/githubintegrations.go:208]: "got some error" or "unexpected error" is enough

File: internal/pkg/plugin/githubintegrations/githubintegrations.go (hunk 3 above)
Comment (lfbdev) [reply internal/pkg/plugin/githubintegrations/githubintegrations.go:1]: will refactor later,  and will abstract out common code between githubaction and githubintegrations at that time

Comment (IronCore864) [reply internal/pkg/plugin/githubintegrations/githubintegrations.go:1]: Some thoughts: trellogh(github), trellogl(gitlab), etc.

File: internal/pkg/plugin/githubintegrations/githubintegrations.go [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+
+	// Note: the file needs to be absent from the repository as you are not
+	// specifying a SHA reference here.
+	opts := &github.RepositoryContentFileOptions{
+		Message: github.String(workflow.commitMessage),
+		Content: []byte(workflow.workflowContent),
+		Branch:  github.String(gi.options.Branch),
+	}
+
+	log.Printf("Creating GitHub Actions workflow %s ...", workflow.workflowFileName)
```
Comment (IronCore864) [inline internal/pkg/plugin/githubintegrations/githubintegrations.go:74]: Logging needs to be updated. not GitHub Actions anymore.

File: internal/pkg/plugin/githubintegrations/githubintegrations.go [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+	)
+
+	// error reason is not 404
+	if err != nil && !strings.Contains(err.Error(), "404") {
+		log.Printf("GetContents failed with error: %s", err)
+		return nil, err
+	}
+	// StatusCode == 404
+	if resp.StatusCode == http.StatusNotFound {
+		log.Printf("GetContents return with status code 404")
```
Comment (IronCore864) [inline internal/pkg/plugin/githubintegrations/githubintegrations.go:148]: returned

File: internal/pkg/plugin/githubintegrations/githubintegrations.go [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+		retMap := mapz.FillMapWithStrAndError(wsFiles, fmt.Errorf("not found"))
+		return retMap, nil
+	}
+	// StatusCode != 200
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("got some error is not expected: %s", resp.Status)
```
Comment (IronCore864) [inline internal/pkg/plugin/githubintegrations/githubintegrations.go:154]: same as the other comment

File: internal/pkg/plugin/githubintegrations/githubintegrations.go [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+	}
+	// StatusCode == 200
+	log.Printf("GetContents return with status code 200")
```
Comment (IronCore864) [inline internal/pkg/plugin/githubintegrations/githubintegrations.go:157]: same

File: internal/pkg/plugin/githubintegrations/githubintegrations.go [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+	var filesInRemoteDir = make([]string, 0)
+	for _, f := range dirContent {
+		log.Printf("Found remote file: %s", f.GetName())
+		filesInRemoteDir = append(filesInRemoteDir, f.GetName())
+	}
+
+	lostFiles := slicez.SliceInSliceStr(wsFiles, filesInRemoteDir)
+	// all files exist
+	if len(lostFiles) == 0 {
+		log.Println("All files exist")
```
Comment (IronCore864) [inline internal/pkg/plugin/githubintegrations/githubintegrations.go:167]: Is "All workflows exist." better?

File: internal/pkg/plugin/githubintegrations/githubintegrations.go [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+		retMap := mapz.FillMapWithStrAndError(wsFiles, nil)
+		return retMap, nil
+	}
+	// some files lost
+	log.Println("Some files lost")
```
Comment (IronCore864) [inline internal/pkg/plugin/githubintegrations/githubintegrations.go:172]: "Some workflow(s) doesn't exist:"
"Workflow: xxx doesn't exist."
"Workflow: yyy doesn't exist."
...

File: internal/pkg/plugin/githubintegrations/githubintegrations.go (hunk 5 above)
Comment (IronCore864) [inline internal/pkg/plugin/githubintegrations/githubintegrations.go:219]: template name needs to be updated. Do a grep over this directory and make sure no "githubactions" should exist.

File: internal/pkg/plugin/githubintegrations/ishealthy.go [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+	retMap, err := gis.VerifyWorkflows(ws)
+	if err != nil {
+		return false, err
+	}
+
+	errFlag := false
+	for name, err := range retMap {
+		if err != nil {
+			errFlag = true
+			log.Printf("The workflow/file %s is not ok: %s", name, err)
```
Comment (IronCore864) [inline internal/pkg/plugin/githubintegrations/ishealthy.go:24]: The logging isn't consistent with verify workflow.

**Review by IronCore864 — APPROVED** [review]

File: cmd/trellogithub/main.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/merico-dev/stream/internal/pkg/plugin/trellogithub"
+)
+
+// NAME is the name of this DevStream plugin.
+const NAME = "trellogithub"
+
+// Plugin is the type used by DevStream core. It's a string.
+type Plugin string
+
+// Install implements the installation of some GitHub Actions workflows.
```
Comment (daniel-hutao) [inline cmd/trellogithub/main.go:15]: GitHub Actions -> trello-github-integ

**Review by daniel-hutao — APPROVED** [review]
lgtm

**Pull request merged**